### PR TITLE
Change the original function arity with decorators

### PR DESCRIPTION
**Pull request by Artpej**
It is not currently possible to change the
the arity of a decorated function.
( see https://gist.github.com/yloiseau/d669c0e30023bf5b5bbc )

Originaly the function `foo`:

```golo
module curry

function curry = |f| -> |a| -> |b| -> f(a, b)

@curry
function foo = |a,b| -> a + b

```

is converted to :

```golo
module curry

function curry = |f| -> |a| -> |b| -> f(a, b)

function foo = |a,b| -> ^curry( |a, b| -> a + b )(a, b)
```

The trick tranforms the original function in a
variable arity function like this :

```golo
module curry

function curry = |f| -> |a| -> |b| -> f(a, b)

function foo = |args...| -> ^curry( |a, b| -> a + b ): invokeWithArguments(args)
```

### CODE REVIEW

**Comment by Artpej** [comment]
Still in progress...

I finally found a simple solution that seems efficient. I need to work a little bit more on this (add some tests, update the documentation, ...) but I think that i will update this PR during the weekend.

**Comment by jponge** [comment]
You rock!

**Comment by Artpej** [comment]
Not enouth time this weekend to complete this PR, but you can already try this new implementation of decorators.

#### under the hood:

previously this code : 
```golo
@decorator
function foo = |a,b| {
  return  a + b
}
```
was translated like this : 
```golo
function foo = |a,b| {
  return decorator( |a,b| -> a + b )(a,b)                                     
}
```
Now the same code is translated like this : 
```golo
function foo = |a,b| {
  return  a + b
}

# the next function is synthetic
function __$$_foo_decorator = |__$$_original| {
  return decorator(__$$_original)                                    
}
```
The `foo` function stay unchanged except it is annotated with `@DecoratedBy("__$$_foo_decorator")`.

At runtime the call of the function `foo`  : 
```golo
function main = |args| {
  foo(12, 30)
}
```
will be resolved mainly like this : 
```golo
function main = |args| {
  __$$_foo_decorator(^foo)(12, 30)
}
```
The example above is just an explaination, the translation is done by `MethodHandle` manipulation with invokedynamic, of course :wink: .

My first benches are very promising (Inlining is very efficient and the JVM seems to be able to do great optimisations with this kind of call).

#### Todo : 

- [ ] Update the documentation (we can now change the arity of the returned function)
- [ ] Update the documentation (explain what's append under the hood), useless?
- [ ] Check that evrything works fine (add some test cases mainly focused on dealing with varargs, also check that decorated function can be passed as a lambda)

**Comment by jponge** [comment]
Lovely :+1:

**Comment by yloiseau** [comment]
Will `foo` still be easilly referencable? Put another way, does `^foo` reference `foo` or `__$$_foo_decorator(^foo)`? Can decorators still cascade? (just some questions that raised, have not yet read the code)

**Comment by Artpej** [comment]
@yloiseau , `^foo` reference return the decorated function, and now the function de-referencment is banged :wink:

**Comment by yloiseau** [comment]
:+1:

**Comment by jponge** [comment]
Bang Bang!

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java [hunk 1]
```diff
@@ -198,65 +199,62 @@ public Object visit(ASTFunctionDeclaration node, Object data) {
...
+      function.addDecorator((Decorator) context.objectStack.pop());
     }
     context.objectStack.push(function);
     if (!function.getDecorators().isEmpty()) {
-      ASTFunction original = (ASTFunction)node.jjtGetChild(0);
-      Node wrapped = wrapFunctionWithDecorators(original, function.getDecorators());
-      wrapped.jjtAccept(this, data);
-    } else {
-      node.childrenAccept(this, data);
+      function.setDecoratorRef("__$$_" + function.getName() + "_decorator_" + context.nextDecoratorId++);
```
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: maybe you should use the decorator name (if available) in the synthetic function name (more readable when debugging). 
Else, I think `__$$_decorated_foo_0` is a better name.

File: src/main/java/fr/insalyon/citi/golo/runtime/AugmentationMethodFinder.java [hunk 2]
```diff
@@ -170,15 +171,19 @@ private boolean matchesArity(Method method) {
...
   private MethodHandle toMethodHandle(Method method) {
-    try {
-      MethodHandle target = lookup.unreflect(method);
-      if (target.isVarargsCollector() && isLastArgumentAnArray(arity, args)) {
-        return target.asFixedArity().asType(type);
-      } else {
-        return target.asType(type);
+    if(isMethodDecorated(method)) {
+      return getDecoratedMethodHandle(lookup, method, arity);
+    }else{
```
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/runtime/AugmentationMethodFinder.java]: add spaces around `else` and after `if` :smile:

File: src/main/java/fr/insalyon/citi/golo/runtime/RegularMethodFinder.java [hunk 3]
```diff
@@ -82,10 +83,14 @@ private MethodHandle toMethodHandle(Method method) throws IllegalAccessException
     if (makeAccessible || isValidPrivateStructAccess(method)) {
       method.setAccessible(true);
     }
-    if ((method.isVarArgs() && isLastArgumentAnArray(type.parameterCount(), args))) {
-      target = lookup.unreflect(method).asFixedArity().asType(type);
-    } else {
-      target = lookup.unreflect(method).asType(type);
+    if(isMethodDecorated(method)) {
+        target = getDecoratedMethodHandle(method, arity);
+    }else{
```
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/runtime/RegularMethodFinder.java]: spaces...

File: src/main/java/fr/insalyon/citi/golo/runtime/AugmentationMethodFinder.java (hunk 2 above)
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/runtime/AugmentationMethodFinder.java]: spaces around else :smile:

**Comment by Artpej** [comment]
@yloiseau do not care about strange names, code formating, and spell checking, it's just an early draft, but working solution :wink:

**Comment by yloiseau** [comment]
@artpej beware that working “prototypes” ends up in production :wink:

File: src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java (hunk 1 above)
Comment (Artpej) [inline src/main/java/fr/insalyon/citi/golo/compiler/ParseTreeToGoloIrVisitor.java]: `__$$_decorated_foo_0` returns the entire decoration, according the function can be decoated by multiple decorators, so I can't use the the decorator name. 

yes `__$$_decorated_foo_0`seems better according this function returns the decorated function.

**Comment by Artpej** [comment]
@yloiseau this things never append in private sectors :innocent: :stuck_out_tongue:

**Comment by Artpej** [comment]
Merge done!
Seems to me ready for , comments, tests, and review...

**Comment by yloiseau** [comment]
:+1:

**Comment by jponge** [comment]
My review is good.

How about... performance? :smile:

**Comment by Artpej** [comment]
it seems that the performances will be improved with the new implementation :+1: 

### 2.1.0 (Before function reference boxing)

```
Result: 1613379,391 ±(99.9%) 15372,852 ops/s [Average]
  Statistics: (min, avg, max) = (1595422,643, 1613379,391, 1628019,921), stdev = 10168,190
  Confidence interval (99.9%): [1598006,539, 1628752,243]


# Run complete. Total time: 00:02:28

Benchmark                                        Mode  Samples          Score          Error  Units
o.g.MyBenchmark.baseline                        thrpt       10  363814951,168 ± 10917256,161  ops/s
o.g.MyBenchmark.decorated_foo                   thrpt       10     243967,133 ±    24380,484  ops/s
o.g.MyBenchmark.decorated_foo_banged            thrpt       10  363852093,943 ± 12000992,130  ops/s
o.g.MyBenchmark.foo                             thrpt       10  360414615,654 ± 14703345,419  ops/s
o.g.MyBenchmark.generic_decorated_foo           thrpt       10     154913,022 ±    18236,608  ops/s
o.g.MyBenchmark.generic_decorated_foo_banged    thrpt       10    1613379,391 ±    15372,852  ops/s
```
I don't really understand why  `decorated_foo_banged` is so efficient, i think is it just because the JVM do some optimisation due to my decorator do nothing. this will never append in the real life.

### 3.0.0-SNAPSHOT (Current master with function reference boxing)

```
Result: 1590192,184 ±(99.9%) 38718,173 ops/s [Average]
  Statistics: (min, avg, max) = (1528942,606, 1590192,184, 1617422,480), stdev = 25609,674
  Confidence interval (99.9%): [1551474,011, 1628910,357]


# Run complete. Total time: 00:02:28

Benchmark                                        Mode  Samples          Score          Error  Units
o.g.MyBenchmark.baseline                        thrpt       10  366998999,881 ±  8079906,399  ops/s
o.g.MyBenchmark.decorated_foo                   thrpt       10     231430,686 ±    28484,385  ops/s
o.g.MyBenchmark.decorated_foo_banged            thrpt       10  367048645,749 ±  9918006,775  ops/s
o.g.MyBenchmark.foo                             thrpt       10  362857443,760 ± 11523806,761  ops/s
o.g.MyBenchmark.generic_decorated_foo           thrpt       10     153856,303 ±    12739,428  ops/s
o.g.MyBenchmark.generic_decorated_foo_banged    thrpt       10    1590192,184 ±    38718,173  ops/s
```
idem for (`decorated_foo_banged`)

### 3.0.0-SNAPSHOT (New decorators implementation)

```
Result: 1584546,192 ±(99.9%) 31007,030 ops/s [Average]
  Statistics: (min, avg, max) = (1558320,964, 1584546,192, 1613495,879), stdev = 20509,230
  Confidence interval (99.9%): [1553539,162, 1615553,221]


# Run complete. Total time: 00:02:28

Benchmark                                        Mode  Samples          Score          Error  Units
o.g.MyBenchmark.baseline                        thrpt       10  364104108,293 ±  8192920,834  ops/s
o.g.MyBenchmark.decorated_foo                   thrpt       10   11137248,958 ±   407877,699  ops/s
o.g.MyBenchmark.decorated_foo_banged            thrpt       10   85086010,842 ±  1911299,393  ops/s
o.g.MyBenchmark.foo                             thrpt       10  358156061,818 ± 25772802,788  ops/s
o.g.MyBenchmark.generic_decorated_foo           thrpt       10     597775,929 ±    14379,550  ops/s
o.g.MyBenchmark.generic_decorated_foo_banged    thrpt       10    1584546,192 ±    31007,030  ops/s
```

## Benches details

```golo
function decorator = |f| -> |a,b| -> f(a, b)

function banged_decorator = |f| -> |a,b| -> f(a, b)

function generic_decorator = |f| -> |args...| -> f: invoke(args)

function generic_banged_decorator = |f| -> |args...| -> f: invoke(args)


function foo = |a,b| -> a + b

@decorator
function decorated_foo = |a,b| -> a + b

@generic_decorator
function generic_decorated_foo = |a,b| -> a + b

@!banged_decorator
function decorated_foo_banged = |a,b| -> a + b

@!generic_banged_decorator
function generic_decorated_foo_banged = |a,b| -> a + b

function foo_bench = |a,b| -> foo(a,b)

function decorated_foo_bench = |a,b| -> decorated_foo(a,b)

function generic_decorated_foo_bench = |a,b| -> generic_decorated_foo(a,b)

function decorated_foo_banged_bench = |a,b| -> decorated_foo_banged(a,b)

function generic_decorated_foo_banged_bench = |a,b| -> generic_decorated_foo_banged(a,b)

```

I juste changed `invoke` by `invokeWithArguments` for the 2.1.0

```java
@State(Scope.Benchmark)
public class MyBenchmark {

  private int x = 42; // not final, just a good mutable field to trick the VM  
  

  @Benchmark
  public Object foo() {
    return MethodsInvocation.foo_bench(x,x);
  }   
  
  @Benchmark
  public Object decorated_foo() {
    return MethodsInvocation.decorated_foo_bench(x,x);
  }  
  
  @Benchmark
  public Object generic_decorated_foo() {
    return MethodsInvocation.generic_decorated_foo_bench(x,x);
  }   
  
  @Benchmark
  public Object decorated_foo_banged() {
    return MethodsInvocation.decorated_foo_banged_bench(x,x);
  }  
  
  @Benchmark
  public Object generic_decorated_foo_banged() {
    return MethodsInvocation.generic_decorated_foo_banged_bench(x,x);
  }     

  @Benchmark
  public Object baseline() {
    return x + x;
  }
}

```

**Comment by jponge** [comment]
Just awesome mate!

Feel-free to merge when you want to.

Thanks for your work!

**Comment by Artpej** [comment]
It seems to me strange that my documentation addition is correct at the first pass :wink:

Just a last question about the `@DecoratedBy` annotation: for instance it take place directly in the `gololang` package, is it the right place? 
Is this thing should be a part of the public API?
In this case, maybe we can create a `gololang.annotations` package for the futur works?

**Comment by jponge** [comment]
> It seems to me strange that my documentation addition is correct at the first pass 

Ok, let me double-check then :-D

> Just a last question about the @DecoratedBy annotation: for instance it take place directly in the gololang package, is it the right place? 
> Is this thing should be a part of the public API?
> In this case, maybe we can create a gololang.annotations package for the futur works?

:+1: for `gololang.annotations` (think about a package-info.java to say that what the package purpose is).

File: doc/decorators.asciidoc [hunk 4]
```diff
@@ -262,6 +262,73 @@ function main = |args| {
...
+
+@curry
+function add = |a,b| -> a + b
+
+function main = |args| {
+  add(12)(30)
+}
+----
+
+The `@curry` decorator transform the `add` function that takes two arguments, in a function
```
Comment (yloiseau) [inline doc/decorators.asciidoc]: into a function

File: doc/decorators.asciidoc [hunk 5]
```diff
@@ -262,6 +262,73 @@ function main = |args| {
...
+that takes only one arguments and return a function that takes the second argument and finally
```
Comment (yloiseau) [inline doc/decorators.asciidoc]: argument and returns

File: src/main/java/fr/insalyon/citi/golo/compiler/ir/GoloFunction.java [hunk 6]
```diff
@@ -98,6 +101,22 @@ public void setSynthetic(boolean synthetic) {
     this.synthetic = synthetic;
   }
 
+  public boolean isDecorated() {
+    return decorated;
+  }
```
Comment (yloiseau) [inline src/main/java/fr/insalyon/citi/golo/compiler/ir/GoloFunction.java]: Why don't you just return `!decorators.isEmpty()` ?

Comment (jponge) [inline src/main/java/fr/insalyon/citi/golo/compiler/ir/GoloFunction.java]: Good catch! 
1 less field is better if you can :-)

File: doc/decorators.asciidoc [hunk 7]
```diff
@@ -262,6 +262,73 @@ function main = |args| {
...
+return the result of the addition.
+
+NOTE: A decorator is applied only if the decorated function is called from Golo
+code. For example if you try to call a decorated Golo function from Java code
+it will be not the decorated function that will be called but the original one.
```
Comment (jponge) [inline doc/decorators.asciidoc]: not be

**Comment by Artpej** [comment]
Done!
Thanks for the review guys!
Ready to merge?

**Comment by jponge** [comment]
For me yes :-)

**Pull request merged**